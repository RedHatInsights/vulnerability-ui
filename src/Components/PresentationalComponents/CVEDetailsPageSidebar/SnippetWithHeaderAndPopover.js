import { Popover, Stack, StackItem } from '@patternfly/react-core';
import propTypes from 'prop-types';
import React from 'react';
import Label from '../Snippets/Label';

const SnippetWithHeaderAndPopover = ({ title, value, content }) => (
    <Popover id="popover" bodyContent={content} headerContent="" position="bottom"
        appendTo={document.querySelector('.vulnerability')}>
        <Stack className="popover-content">
            <StackItem>
                <Label isLarge>{title}</Label>
            </StackItem>
            <StackItem>
                <a>{value}</a>
            </StackItem>
        </Stack>
    </Popover>
);

SnippetWithHeaderAndPopover.propTypes = {
    title: propTypes.any,
    value: propTypes.any,
    content: propTypes.any
};

export default SnippetWithHeaderAndPopover;
