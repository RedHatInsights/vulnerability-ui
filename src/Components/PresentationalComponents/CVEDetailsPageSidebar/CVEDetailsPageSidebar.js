import { Stack, StackItem, Split, SplitItem, TextContent } from '@patternfly/react-core';
import parseCvssScore from '@redhat-cloud-services/frontend-components-utilities/parseCvssScore';
import propTypes from 'prop-types';
import React from 'react';
import { getImpactDetails } from '../../../Helpers/MiscHelper';
import CvssVector from '../CvssVector/CvssVector';
import messages from '../../../Messages';
import SnippetWithHeaderAndPopover from './SnippetWithHeaderAndPopover';
import { ExclamationTriangleIcon } from '@patternfly/react-icons';
import { Table, TableBody, TableHeader } from '@patternfly/react-table';
import { STATUS_OPTIONS } from '../../../Helpers/constants';
import { injectIntl } from 'react-intl';
import { Shield } from '@redhat-cloud-services/frontend-components/Shield';
import Label from '../Snippets/Label';
import { CVEPageContext } from '../../SmartComponents/CVEDetailsPage/CVEDetailsPage';
import WithLoader from '../WithLoader/WithLoader';

const CVEDetailsPageSidebar = ({ cveAttributes, intl }) => {
    const cveDetails = getImpactDetails(cveAttributes.impact || 'Unknown');

    const {
        business_risk: businessRisk,
        business_risk_justification: businessRiskJustification,
        status,
        status_justification: statusJustification,
        systems_status_detail: systemsStatusDetail,
        systems_status_divergent: systemsStatusDivergent
    } = cveAttributes;

    const detailRows =
        systemsStatusDetail &&
        Object.keys(systemsStatusDetail).map(item => [
            STATUS_OPTIONS.find(status => status.value === item.toString()).label,
            systemsStatusDetail[item].toString()
        ]);

    const brPopoverContent = (
        <Stack className="pf-u-p-xs">
            <StackItem>
                <Label className="pf-u-mb-xs">
                    {intl.formatMessage(messages.justificationNote)}
                </Label>
            </StackItem>
            <StackItem>{businessRiskJustification || '--'}</StackItem>
        </Stack>
    );

    const statusPopoverContent = (
        <Stack className="pf-u-p-xs">
            <StackItem>
                <Label className="pf-u-mb-xs">
                    {intl.formatMessage(messages.cveStatus)}
                </Label>
                {status || '--'}
                <Label className="pf-u-mb-xs pf-u-mt-sm">
                    {intl.formatMessage(messages.justificationNote)}
                </Label>
                {statusJustification || '--'}
            </StackItem>
            {systemsStatusDivergent > 0 && (
                <StackItem>
                    <TextContent>
                        <hr className="pf-u-mt-sm pf-u-mb-sm" />
                    </TextContent>
                    <Label isLarge>
                        {intl.formatMessage(messages.cveDetailsPageSystemsWTCves)}
                    </Label>
                    <Table
                        aria-label="Status details"
                        gridBreakPoint=""
                        cells={[intl.formatMessage(messages.statusLabel), intl.formatMessage(messages.count)]}
                        rows={detailRows}
                        className="status-breakdown-table"
                        variant="compact"
                    >
                        <TableHeader />
                        <TableBody />
                    </Table>
                </StackItem>
            )}
        </Stack>
    );

    return (
        <CVEPageContext.Consumer>
            {context => (
                <Stack hasGutter>
                    <StackItem>
                        <Split hasGutter>
                            <SplitItem>
                                <SnippetWithHeaderAndPopover
                                    title={intl.formatMessage(messages.businessRiskLabel)}
                                    value={
                                        <WithLoader isLoading={context.isLoading} style={{ width: '100px' }}>
                                            {businessRisk}
                                        </WithLoader>}
                                    content={brPopoverContent}
                                />
                            </SplitItem>
                            <SplitItem className="pf-u-ml-lg">
                                <SnippetWithHeaderAndPopover
                                    title={intl.formatMessage(messages.statusLabel)}
                                    content={statusPopoverContent}
                                    value={
                                        <WithLoader isLoading={context.isLoading} style={{ width: '100px' }}>
                                            <span>
                                                {systemsStatusDivergent > 0 && (
                                                    <ExclamationTriangleIcon color="var(--pf-global--primary-color--100)" />
                                                )}{' '}
                                                {status}
                                            </span>
                                        </WithLoader>
                                    }
                                />
                            </SplitItem>
                        </Split>
                    </StackItem>

                    <StackItem>
                        <Label className="pf-u-mb-xs" isLarge>
                            {intl.formatMessage(messages.impact)}
                        </Label>
                        <WithLoader isLoading={context.isLoading} style={{ width: '100px' }}>
                            <span id="severity-shield" style={{ color: cveDetails.color }}>
                                <Shield impact={cveDetails.title} hasLabel />
                            </span>
                        </WithLoader>
                    </StackItem>

                    <StackItem>
                        <CvssVector
                            cvss2_metrics={cveAttributes.cvss2_metrics}
                            cvss3_metrics={cveAttributes.cvss3_metrics}
                            score={parseCvssScore(cveAttributes.cvss2_score, cveAttributes.cvss3_score)}
                        />
                    </StackItem>
                </Stack>
            )}
        </CVEPageContext.Consumer>
    );
};

CVEDetailsPageSidebar.propTypes = {
    cveAttributes: propTypes.any,
    intl: propTypes.any
};

export default injectIntl(CVEDetailsPageSidebar);
