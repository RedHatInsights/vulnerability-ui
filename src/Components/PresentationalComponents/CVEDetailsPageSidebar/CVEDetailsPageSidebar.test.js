import React from 'react';
import CVEDetailsPageSidebar from './CVEDetailsPageSidebar';
import { render } from '@testing-library/react';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

const cveAttributes = {
    business_risk: "Low",
    business_risk_id: 1,
    business_risk_justification: "m",
    cvss2_metrics: null,
    cvss2_score: null,
    cvss3_metrics: "CVSS:3.1/AV:L/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H",
    cvss3_score: "7.800",
    description: "A flaw was found in sudo. A heap-based buffer overflow was found in the way sudo parses command line arguments. This flaw is exploitable by any local user who can execute the sudo command (by default, any local user can execute sudo) without authentication. Successful exploitation of this flaw could lead to privilege escalation. The highest threat from this vulnerability is to data confidentiality and integrity as well as system availability.",
    impact: "Important",
    public_date: "26 Jan 2021",
    status: "Resolved",
    status_id: 4,
    status_justification: null,
    synopsis: "CVE-2021-3156",
    systems_status_divergent: 206,
    systems_status_detail: {
        0: 114,
        4: 4
    }
}

describe('CVEDetailsPageSidebar component', () => {
    it.skip('should render correctly', () => {
        const { asFragment } = render(
            <IntlProvider locale="en">
                <CVEDetailsPageSidebar cveAttributes={cveAttributes} />
            </IntlProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
