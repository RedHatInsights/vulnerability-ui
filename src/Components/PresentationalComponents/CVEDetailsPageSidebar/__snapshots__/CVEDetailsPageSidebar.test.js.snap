// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVEDetailsPageSidebar component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-l-stack pf-m-gutter"
  >
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-l-split pf-m-gutter"
      >
        <div
          class="pf-v5-l-split__item"
        >
          <div
            style="display: contents;"
          >
            <div
              class="pf-v5-l-stack popover-content"
            >
              <div
                class="pf-v5-l-stack__item"
              >
                <span
                  class="vuln-label "
                  style="font-size: medium; display: block;"
                >
                  Business risk
                </span>
              </div>
              <div
                class="pf-v5-l-stack__item"
              >
                <a>
                  Low
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-v5-l-split__item pf-v5-u-ml-lg"
        >
          <div
            style="display: contents;"
          >
            <div
              class="pf-v5-l-stack popover-content"
            >
              <div
                class="pf-v5-l-stack__item"
              >
                <span
                  class="vuln-label "
                  style="font-size: medium; display: block;"
                >
                  Status
                </span>
              </div>
              <div
                class="pf-v5-l-stack__item"
              >
                <a>
                  <span>
                    <span
                      class="pf-v5-c-icon"
                    >
                      <span
                        class="pf-v5-c-icon__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg pf-v5-u-mr-xs"
                          color="var(--pf-v5-global--primary-color--100)"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 576 512"
                          width="1em"
                        >
                          <path
                            d="M569.517 440.013C587.975 472.007 564.806 512 527.94 512H48.054c-36.937 0-59.999-40.055-41.577-71.987L246.423 23.985c18.467-32.009 64.72-31.951 83.154 0l239.94 416.028zM288 354c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
                          />
                        </svg>
                      </span>
                    </span>
                     Resolved
                  </span>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <span
        class="vuln-label pf-v5-u-mb-xs"
        style="font-size: medium; display: block;"
      >
        Severity
      </span>
      <span
        id="severity-shield"
      >
        <span>
          <div
            style="display: contents;"
          >
            <span>
              <span
                class="pf-v5-c-icon pf-m-md"
              >
                <span
                  class="pf-v5-c-icon__content"
                >
                  <svg
                    aria-hidden="false"
                    aria-label="Important"
                    class="pf-v5-svg"
                    color="#ec7a08"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    viewBox="0 0 896 1024"
                    width="1em"
                  >
                    <path
                      d="M861.5,0 L34.5,0 C15.4,0 0,14.3 0,32 L0,452.1 C0,768 387.7,1024 448.5,1024 C509.3,1024 896,768 896,452.2 L896,32 C896,14.3 880.6,0 861.5,0 Z M490.7,768 L405.3,768 C393.5,767.8 384.2,757.5 384,744.7 L384,663.3 C384.2,650.5 393.6,640.3 405.3,640 L490.7,640 C502.5,640.2 511.8,650.5 512,663.3 L512,744.7 L512.1,744.7 C511.8,757.5 502.4,767.8 490.7,768 Z M543.9,162.7 L517.2,514.4 C515.8,530.9 502,544 485.3,544 L410.6,544 C394,544 380.1,531.2 378.7,514.7 L352.1,163 C350.5,144.3 365.3,128.3 384,128.3 L512,128 C530.7,128 545.4,144 543.9,162.7 Z"
                    />
                  </svg>
                </span>
              </span>
               Important
            </span>
          </div>
        </span>
      </span>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-content"
      >
        <div
          style="display: contents;"
        >
          <span
            class="vuln-label pf-v5-u-mb-xs pointer"
            style="font-size: medium; display: block;"
          >
            CVSS 3.0  base score
            <span
              class="pf-v5-c-icon"
            >
              <span
                class="pf-v5-c-icon__content"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg pf-v5-u-ml-xs"
                  color="var(--pf-v5-global--secondary-color--100)"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 448c-110.532 0-200-89.431-200-200 0-110.495 89.472-200 200-200 110.491 0 200 89.471 200 200 0 110.53-89.431 200-200 200zm107.244-255.2c0 67.052-72.421 68.084-72.421 92.863V300c0 6.627-5.373 12-12 12h-45.647c-6.627 0-12-5.373-12-12v-8.659c0-35.745 27.1-50.034 47.579-61.516 17.561-9.845 28.324-16.541 28.324-29.579 0-17.246-21.999-28.693-39.784-28.693-23.189 0-33.894 10.977-48.942 29.969-4.057 5.12-11.46 6.071-16.666 2.124l-27.824-21.098c-5.107-3.872-6.251-11.066-2.644-16.363C184.846 131.491 214.94 112 261.794 112c49.071 0 101.45 38.304 101.45 88.8zM298 368c0 23.159-18.841 42-42 42s-42-18.841-42-42 18.841-42 42-42 42 18.841 42 42z"
                  />
                </svg>
              </span>
            </span>
          </span>
        </div>
        <span
          class="pf-v5-u-mr-md"
        >
          7.8
        </span>
        <span
          id="cvss-vector-content"
        >
          Vector: AV:L/AC:L/PR:L/UI:N/S:U/C:H/I:H/A:H
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
